templates/year/Makefile.inc

```diff
@@ -1,500 +1,192 @@
+	$(call find-graphics-deps,$(1),$(2),plt,tex eps)\
+	$(call find-graphics-deps,$(1),$(2),plt,tex eps,color/)\
+	$(call find-graphics-deps,$(1),$(2),ipe,pdf svg png)\
+	$(call find-graphics-deps,$(1),$(2),svg,pdf svg png)\
+	$(call find-graphics-deps,$(1),$(2),eps,pdf svg png)\
+	$(call find-graphics-deps,$(1),$(2),pdf,pdf svg png)\
+	$(call find-graphics-deps,$(1),$(2),png,png)\
+	$(call find-graphics-deps,$(1),$(2),jpg,jpg)\
+	$(call find-graphics-deps,$(1),$(2),jpeg,jpeg)
+
+
+# == Data retrieval ==
+GLOB_DATA := ../data
+# credentials are at least obfuscated so they aren't easily grep'able from logs
+SOAP_REQUEST = $(SOAPCLIENT) $(WS_URL) $(1) $(2) $(shell echo -e $(WS_USERNAME) | base64) $(shell echo -e $(WS_PASSWORD) | base64)
+
+# vysledky?-% is implicit rule, results can't be also imlicit, so mention it explicitly
+# see: https://www.gnu.org/software/make/manual/html_node/Chained-Rules.html
+$(GLOB_DATA)/results.xml: $(GLOB_DATA)/resultsRequest.soap
+$(GLOB_DATA)/stats.xml: $(GLOB_DATA)/statsRequest.soap
+
+$(GLOB_DATA)/%.xml: $(GLOB_DATA)/%Request.soap
+	$(call SOAP_REQUEST,$<,$@)
+
+.PRECIOUS: $(GLOB_DATA)/*.xml $(GLOB_DATA)/results.xml
+
+
+# == Data processing ==
+SIGNATURES := ../signatures.tex
+$(SIGNATURES): $(GLOB_DATA)/signatures.xml $(SIGNATURES_STYLE)
+	xsltproc $(XSL_SIGNATURES) $< >$@
+
+$(RESULTS)/vysledky%.tex: $(GLOB_DATA)/results.xml $(XSL_RESULTS)
+	xsltproc --stringparam type brojure \
+		--stringparam category $(word 2,$(subst -, ,$(subst .tex,,$(subst $(RESULTS)/vysledky,,$@)))) \
+		--stringparam series   $(word 1,$(subst -, ,$(subst .tex,,$(subst $(RESULTS)/vysledky,,$@)))) \
+		$(XSL_RESULTS) $< >$@
+
+$(RESULTS)/statistiky%.csv: $(GLOB_DATA)/stats.xml $(XSL_STATS)
+	xsltproc --stringparam series $* \
+		$(XSL_STATS) $< >$@
+	# apply created results to statistics
+	$(STATS_SH) $(PROBLEMS) $@
+
+$(RESULTS)/%.tex: $(RESULTS)/%.csv
+	iconv -f latin2 -t utf8 $< | \
+	cut -d\; -f1,2,3,4,5,6,7,8,9,10,12,13,14 | \
+	sed '/^$$/d;/^%/d;s/;;/;--;/g;s/;;/;--;/g;s/;/ \& /g;s/$$/\\\\/' | sed -e '2i\\\\midrule'  > $@
+
+
+# == Deadlines ==
+../Deadline.inc: ../Deadline_xml.inc
+	echo "#" > $@
+	echo "# Automaticky generovaný" >> $@
+	sed -e 's/DEADLINE_XML\[\([0-9]*\)\]=\([0-9]*\)-[0]*\([1-9][0-9]*\)-[0]*\([1-9][0-9]*\)T[0]*\([1-9][0-9]*\):\([0-9]*\):\([0-9]*\)/DEADLINE\[\1\]=\4.~\3.~\2 \5.\6/;s/DEADLINES_XML\[\([0-9]*\)\]=\([0-9]*\)-[0]*\([1-9][0-9]*\)-[0]*\([1-9][0-9]*\)/DEADLINES\[\1\]=\4.~\3.~\2/' $< >> $@
+
+# == Makefile configuration ==
+../Makefile.conf:
+	cp ../Makefile.conf.sample ../Makefile.conf
+
+# == TEX to PDF compiler ==
+define XELATEX
+	xelatex -output-directory $(@D) -jobname '$(basename $(@F))' -no-pdf $(1) || (rm -f $@ && false)
+	xelatex -output-directory $(@D) -jobname '$(basename $(@F))' $(1) || (rm -f $@ && false)
+endef
+
+# == Author solutions ==
+.SECONDEXPANSION:
+$(OUT)/reseni%.pdf: $(PROBLEMS)/problem%.tex $$(call graphics-deps,$(PGRAPHICS),$$*) $(SIGNATURES)

```

Comment:
Proč je tady ten dvojdolar před `call`?
Použití `$$*` v tomhle kontextu funguje? (Čekal bych jen `$*`, ale i tak mě překvapuje, že proměnná je již definovaná v závislosti.)

Reply:
Při překladu jednotlivých úloh chceme najít obrázky, které mají stejný číslo úlohy. Toho docílíme tak, že číslo úlohy dáme jako argument do funkce `graphics-deps`. Nicméně jak si sám řekl, proměnná `$*` není v normálním případě ještě definovaná a tedy by nemohla být použita. Proto se zde využívá triku [sekundární expanze](https://www.gnu.org/software/make/manual/html_node/Secondary-Expansion.html), kdy při první expanzi se definuje proměnná `$*` a až při druhé se použije v `call`, který vypíše všechny potřebné obrázky. To je taky důvod, proč `call` má před sebou dvojdolar. Chceme totiž, aby se ta funkce spustila až po definici té proměnné.

Celé toto je taky důvod, proč je před tímto pravidlem `.SECONDEXPANSION:`, protože tím maku říkáme, že chceme této funkcionality využít.